email/parser.go

```diff
@@ -74,146 +77,179 @@ func (p *Parser) Stop() error {
 	}
 }
 
-// threadedParseMessages will periodically fetch email messages that have not
-// been parsed yet and parse them.
-func (p *Parser) threadedParseMessages() {
+// buildAbuseReport will parse the email body into an abuse report. This report
+// contains information about the reporter, the tags and the skylinks.
+func (p *Parser) buildAbuseReport(email database.AbuseEmail) (database.AbuseReport, error) {
 	// convenience variables
-	abuseDB := p.staticDatabase
 	logger := p.staticLogger
 
-	ticker := time.NewTicker(parseFrequency)
-	first := true
+	// check for nil body
+	body := email.Body
+	if body == nil {
+		return database.AbuseReport{}, errors.New("empty body")
+	}
 
-	// start the loop
-	for {
-		// sleep until next iteration, sleeping at the start of the for loop
-		// allows to 'continue' on error.
-		if !first {
-			select {
-			case <-p.staticContext.Done():
-				logger.Info("Parser context done")
-				return
-			case <-ticker.C:
-			}
-		}
-		first = false
+	// extract the reporter.
+	reporter := database.AbuseReporter{
+		Email: email.From,
+	}
 
-		logger.Debugln("Triggered")
+	// extract all tags and skylinks
+	var tags []string
+	var skylinks []string
 
-		// fetch all unparsed emails
-		toParse, err := abuseDB.FindUnparsed()
-		if err != nil {
-			logger.Errorf("Failed fetching unparsed emails, error %v", err)
-			return
-		}
+	// use the message library to parse the email
+	msg, err := message.Read(bytes.NewBuffer(body))
+	if err != nil {
+		return database.AbuseReport{}, err
+	}
 
-		// log unparsed messages count
-		numUnparsed := len(toParse)
-		if numUnparsed == 0 {
-			logger.Debugf("Found %v unparsed messages\n", numUnparsed)
-			continue
-		}
-		logger.Infof("Found %v unparsed messages\n", numUnparsed)
+	// if the message is multi-part, read every part and only parse the ones
+	// with Content-Type 'text/plain'
+	mpr := msg.MultipartReader()
+	if mpr != nil {
+		for {
+			p, err := mpr.NextPart()
+			if err == io.EOF {
+				break
+			} else if err != nil {
+				logger.Errorf("error occurred while trying to read next part from multi-part reader, err: %v", err)
+				break
+			}
 
-		// loop all emails and parse them
-		for _, email := range toParse {
-			err = func() (err error) {
-				lock := abuseDB.NewLock(email.UID)
-				err = lock.Lock()
-				if err != nil {
-					return errors.AddContext(err, "could not acquire lock")
-				}
-				defer func() {
-					unlockErr := lock.Unlock()
-					if unlockErr != nil {
-						err = errors.Compose(err, errors.AddContext(unlockErr, "could not release lock"))
-						return
-					}
-				}()
-
-				// parse the email
-				report, err := p.parseEmail(email)
+			t, _, _ := p.Header.ContentType()
+			if t == "text/plain" {
+				body, err = ioutil.ReadAll(p.Body)
```

Comment:
Can't you use the reader directly? That would be more efficient.

Also, we should probably make sure incoming emails don't exceed a certain size.

Reply:
We need to extract both the tags and the skylinks, so I'd need to read from the reader multiple times. I know we could use a `TeeReader` but I don't see why we can't just read the entire body and pass around byte slices.

Reply:
I capped the email body at 8MiB for now.  I don't error out though if it exceeds that, maybe I should? Can F/U with that.

Reply:
I chose 8 MiB because that's well below the 16MiB mongo document limit, I guess we can lower that to a way lower number, like kilobytes but at the same time, I think "just" capping it is sufficient.